### PR TITLE
optimize perfermance of fd_read/fd_write

### PR DESCRIPTION
**Pull request by liutao-liu**
- Issue #7973

Compared to native(C/C++), ```fd-read``` can only reach 20% of native. I will divide the performance hotspots  into the following two categories:,
1. Wasi implementation of ```fd-read```(approximately 30% of the hotspots), such as the cost of [redundant memory copies,](https://github.com/bytecodealliance/wasmtime/blob/86dcae460b5437703570e52253cc1b7d5bad8597/crates/wasi/src/preview1.rs#L1602) the cost of [inserting](https://github.com/bytecodealliance/wasmtime/blob/86dcae460b5437703570e52253cc1b7d5bad8597/crates/wasi/src/preview1.rs#L1569) and [deleting](https://github.com/bytecodealliance/wasmtime/blob/86dcae460b5437703570e52253cc1b7d5bad8597/crates/wasi/src/preview1.rs#L1577) ```HostInputStream``` into ```WasiView```, and the cost of accessing guest memory(merged [PR);](https://github.com/bytecodealliance/wasmtime/pull/8277)
2. Wasm program call host function(approximately 40% of the hotspots): I am not very familiar with this mechanism yet, and currently I cannot understand the reasons behind the hotspots here;

This PR is mainly optimized for the first type of performance hotspot:
1. Avoid unnecessary copying;
2. Avoid inserting and deleting ```HostInputStream``` into```WasiView```;

The perfermance hotspots of ```fd_write``` is the same ```fd_read```:
https://github.com/bytecodealliance/wasmtime/blob/86dcae460b5437703570e52253cc1b7d5bad8597/crates/wasi/src/preview1.rs#L1682

https://github.com/bytecodealliance/wasmtime/blob/86dcae460b5437703570e52253cc1b7d5bad8597/crates/wasi/src/preview1.rs#L244

### CODE REVIEW

File: crates/wasi/src/preview1.rs [hunk 1]
```diff
@@ -1550,37 +1592,39 @@ impl wasi_snapshot_preview1::WasiSnapshotPreview1 for WasiP1Ctx {
...
-                let read = read?;
-                let n = read.len().try_into()?;
-                let pos = pos.checked_add(n).ok_or(types::Errno::Overflow)?;
+
+                let file = self.table().get(&fd)?.file()?;
+                let read_size = file
+                    .file
+                    .read_at(buf, pos)
+                    .or_else(|e| Err(StreamError::LastOperationFailed(e.into())))?;
+                buf.to_vec().truncate(read_size);
```
Comment (alexcrichton) [inline crates/wasi/src/preview1.rs:1619]: It's ok to omit this line, this isn't actually doing anything. As-written this is copying data from the guest to the host, truncating it, and then deallocating it.

File: crates/wasi/src/preview1.rs [hunk 2]
```diff
@@ -276,6 +277,47 @@ impl BlockingMode {
...
+        append: bool,
+        mut pos: u64,
+    ) -> Result<usize> {
+        let buf = bytes.as_slice()?.expect("got none");
+        let mut buf = buf.deref();
+
+        let mut total: usize = 0;
+        while !buf.is_empty() {
+            // NOTE: should check FILE_WRITE_CAPACITY
+            let len = buf.len().min(4096usize);
```
[the remark this answers is not in the document]
Comment (alexcrichton) [reply crates/wasi/src/preview1.rs:294]: Also given the inner loops below the outer loop can be removed too.

File: crates/wasi/src/preview1.rs [hunk 3]
```diff
@@ -1550,37 +1592,39 @@ impl wasi_snapshot_preview1::WasiSnapshotPreview1 for WasiP1Ctx {
...
             }) => {
                 let fd = fd.borrowed();
-                let blocking_mode = *blocking_mode;
                 let position = position.clone();
                 drop(t);
                 let Some(buf) = first_non_empty_iovec(iovs)? else {
                     return Ok(0);
                 };
 
+                let mut buf = buf.as_slice_mut()?.expect("get slice error");
```
[the remark this answers is not in the document]
Comment (liutao-liu) [reply crates/wasi/src/preview1.rs:1609]: got

File: crates/wasi/src/preview1.rs (hunk 1 above)
Comment (liutao-liu) [reply crates/wasi/src/preview1.rs:1619]: got

File: crates/wasi/src/preview1.rs (hunk 2 above)
[the remark this answers is not in the document]
Comment (liutao-liu) [reply crates/wasi/src/preview1.rs]: got

**Comment by liutao-liu** [comment]
@alexcrichton  Your review comments have been completed，this pr can be merged.

**Pull request merged**